Fix for error "An item with the same key has already been added."  and AspNet Test Server Project

alexbclarke opened the pull request:
Hi Austin,
Thanks for sharing your work!
I made a couple of changes to the solution. One to avoid an error "An item with the same key has already been added." when a Json Rpc service class is instantiated from the Global.asax.cs Init() method where it can be called multiple times. Another change is, I added a TestServer_AspNet project that can be used to reproduce the error and outline possible solutions.

Hope you will find these changes useful.
Thanks again,

Best regards,
Alexander.

alexbclarke committed:
Fix: Error "An item with the same key has already been added." when a JSON RPC service class in instantiated in the Global.asax.cs Init() method where it can be called several times.

alexbclarke committed:
Add TestServer_AspNet project with sample code using WebActivator to instantiate Json Rpc services

alexbclarke committed:
Downgrade TestServer_AspNet .Net framework to 4.0 (attempting to fix build error)

alexbclarke committed:
Add NUnit3TestAdapter 3.7.0 and update NUnit to 3.7.1 to allow running tests in Visual Studio

Astn commented:
Thanks for the contribution. However, I am not seeing the tests you refer to in this merge request.

alexbclarke committed:
Reproduce "An item with the same key has already been added" error when instantiating the same service more than once. Add TestServer_AspNet to demonstrate that Global.Init() is called several times and instantiating a service there can cause the error. Suggested fix is commented out. Add NUnit3TestAdapter to allow running test from Visual Studio.

alexbclarke commented:
Hi Austin, there was no unit test. Initially I added Test_AspNet web project that demonstrates the error. However, Travis would fail on build, so I removed that project. I now added Test_AspNet project again and added a unit test. Please review and let me know if my changes could add value to your project.

alexbclarke committed:
Fix "An item with the same key has already been added" error when instantiating the same service more than once.

Astn commented on AustinHarris.JsonRpcTestN/Test.cs:
```diff
@@ -1879,6 +1879,13 @@ public void TestNestedReturnType()
+        public void InstantiatingAServiceMoreThanOnceDoesNotThrow()
+        {
+            Assert.DoesNotThrow(()=>new CalculatorService());
+            Assert.DoesNotThrow(() => new CalculatorService());
```
I am concerned that by allowing this behavior that we would be introducing risk that a service does not behave as expected. This this example we have created two different instances of CalculatorService which may contain different internal state. The way this is being done, currently only the first service will be used to handle requests.
Alternatively a context can be added when registering a service to allow multiple instances to handle requests, each for their own context.

Astn replied:
So the reason it is failing is that two different instances are being bound to the same SessionId. If two instances are desired, then they need different sessionIds

Astn replied:
I think those _workarounds_ are fantastic! We should add them to the getting started documentation. I currently believe that it is the expected and correct behavior to throw the exception when trying to register a handler for the same call twice in the same scope. We could possible make it a little nicer by having additional ways to register services that will only do it if it's needed, or let you provide a delegate to create the service.

The reason I am still feeling this way is because throwing the exception makes it very clear to the developer that we could not register their handler and that it failed.

Based off of this test (https://github.com/Astn/JSON-RPC.NET/blob/master/AustinHarris.JsonRpcTestN/Test.cs#L63)
See this contrived scenario:
```
        [Test()]
        public void TestCanCreateAndRemoveSession()
        {
            var h = JsonRpc.Handler.GetSessionHandler("this one");
            var metadata = new System.Collections.Generic.List<Tuple<string, Type>> {
                Tuple.Create ("fruit", typeof(string)),
                Tuple.Create ("returns", typeof(string))
            }.ToDictionary(x => x.Item1, x => x.Item2);

// First Registration - A function that returns "Red - $fruit" is bound to the "getColor" function
            h.RegisterFuction("getColor", metadata, new System.Collections.Generic.Dictionary<string, object>(),new Func<string, string>(x => "Red -" + x));

// Duplicate Registration (Throws Error under current code) - A function that returns "Blue - $fruit" is bound to the "getColor" function. This will never fire if "getColor" is already bound (which it is). We currently throw an error under this condition. Your patch would prevent the error, but the test would still fail. Maybe we should make the error message to be better? If we swallow the error, then the developer will not know that this binding failed.
            h.RegisterFuction("getColor", metadata, new System.Collections.Generic.Dictionary<string, object>(),new Func<string, string>(x => "Blue -" + x));

            string request = @"{method:'getColor',params:{'fruit':'Bananna'},id:1}";
            string expectedResult = "{\"jsonrpc\":\"2.0\",\"result\":\"Red - Bananna\",\"id\":1}";
            string expectedResultAfterDestroy = "{\"jsonrpc\":\"2.0\",\"error\":{\"message\":\"Method not found\",\"code\":-32601,\"data\":\"The method does not exist / is not available.\"},\"id\":1}";
            var result = JsonRpcProcessor.Process("this one", request);
            result.Wait();

            var actual1 = JObject.Parse(result.Result);
            var expected1 = JObject.Parse(expectedResult);
            Assert.AreEqual(expected1, actual1);

            h.Destroy();

            var result2 = JsonRpcProcessor.Process("this one", request);
            result2.Wait();

            Assert.AreEqual(JObject.Parse(expectedResultAfterDestroy), JObject.Parse(result2.Result));
        }
```

What do you think?

alexbclarke replied:
Glad you found the workarounds useful :)

I think you are right. Having an exception makes it cleaner and less hidden when something is not acceptable. What I would like to suggest is to make the error message clearer than a generic "An item with the same key ...". 
Have a look at this commit ["Make error message clearer when registering methods that have the same name"](https://github.com/alexbclarke/JSON-RPC.NET/commit/7c98955975d5a210a2d5c375c69fc2ef4db5e6eb)

I have also fixed a few tests failing simply because the expected result string was slightly miss-formatted: [Fix tests failing due to miss-formatted expected results.](https://github.com/alexbclarke/JSON-RPC.NET/commit/94b8aa1372063283f36c415dde23ef9b881f69a6)

Astn replied:
I've added a few comments on those commits you referenced. :+1:

alexbclarke committed:
Make error message in clearer when registering methods that have the same name.

alexbclarke committed:
Fix tests failing due to miss-formatted expected results.

alexbclarke committed:
Downgrade NUnit to 3.2.1 to prevent failing of Assert.AreEqual() on objects where properties only differ by position. Revert "Fix tests failing due to miss-formatted expected results" commit 94b8aa1372063283f36c415dde23ef9b881f69a6. Add ThrowArgumentException_WhenInstantiatingAServiceMoreThanOnce() test.

razvancurcudel commented:
Hi. Will this branch be merged soon ? The error is still there :(

Astn committed:
Merge branch 'master' into master

Astn commented:
@razvancurcudel this PR adds some tests and enhances the exception that is thrown when your code tries to register the same service multiple times. There are some build fixes that need to happen before I can merge this.
After it is merged you will still get an exception with trying to register a service multiple times with the same service name. It will just be a better message in the exception.

razvancurcudel commented:
What is the workaround for this problem? I started an empty asp.net project and I added a service with a rpc method as in example. I initialised that service in Init method from Global.asax class and I got the error.
I moved the initialization into Application_start method and it seems to work now, but I don't know if it is the right thing to do.

Astn commented:
Anything that is only called once would be perfect!
Sounds like we need to change the wiki to say App_Start instead of Init.

Astn commented:
Workaround here: https://github.com/Astn/JSON-RPC.NET/pull/83#discussion_r124734273
